### ISSUE TITLE
Incorrect output when assembling a form with TensorElement and symmetry=True

### ISSUE THREAD

**Issue by pbrubeck** [firedrakeproject/firedrake#1955]
Consider the following snippet that assembles the simplest linear form on a symmetric tensor function space:
```python
from firedrake import *
mesh = UnitSquareMesh(1,1,quadrilateral=False)
Q = TensorFunctionSpace(mesh, "DG", 0, symmetry=True)
q = TestFunction(Q)
I = Identity(2)
Bq = assemble(inner(q,I)*dx)
print(Bq.dat.data_ro)
```
The form does not depend on the off-diagonal entries of `q`, and hence the use symmetry is valid.
The correct output should be a scaled identity on the two nodes, but instead it comes in a scrambled fashion,
```
[[[0.5 0. ]
  [0.5 0. ]]
 [[0.5 0. ]
  [0.5 0. ]]]
```

For `degree=1` we get
```
[[[0.16666667 0.        ]
  [0.16666667 0.16666667]]

 [[0.         0.16666667]
  [0.16666667 0.        ]]

 [[0.16666667 0.        ]
  [0.         0.        ]]

 [[0.16666667 0.        ]
  [0.16666667 0.16666667]]

 [[0.         0.16666667]
  [0.16666667 0.        ]]

 [[0.16666667 0.        ]
  [0.         0.        ]]]
```

**Comment by wence-** [comment]
I think that the kernels are right, but the global datastructures that Firedrake makes are wrong. For a tensor element with symmetry. `tsfc.finatinterface.create_element` produces

```
with_symmetry.space_dimension() => 3
without_symmetry.space_dimensino() => 4
```

But 

```
In [41]: mesh = UnitTriangleMesh()                                                                                

In [42]: V = TensorFunctionSpace(mesh, "DG", 0, symmetry=True)                                                    

In [43]: f = Function(V)                                                                                          

In [44]: V.finat_element.space_dimension()                                                                        
Out[44]: 1

In [45]: f.dat.cdim                                                                                               
Out[45]: 4
```

This is because we make something with as many components as the `value_shape()` of the tensor element. But for elements with symmetry, it should be as many components as the `reference_value_shape()`. This patch "fixes" things, but I am not sure if there are any further consequences

```patch
diff --git a/firedrake/functionspaceimpl.py b/firedrake/functionspaceimpl.py
index 88553e87..9317a873 100644
--- a/firedrake/functionspaceimpl.py
+++ b/firedrake/functionspaceimpl.py
@@ -287,7 +287,7 @@ class FunctionSpace(object):
         if type(element) is ufl.TensorElement:
             # UFL enforces value_shape of the subelement to be empty
             # on a TensorElement.
-            self.shape = element.value_shape()
+            self.shape = element.reference_value_shape()
         elif type(element) is ufl.VectorElement:
             # First dimension of the value_shape is the VectorElement
             # shape.
```

```
In [2]: mesh = UnitSquareMesh(1, 1)                                                                               

In [3]: V = TensorFunctionSpace(mesh, "DG", 0, symmetry=True)                                                     

In [4]: f = TestFunction(V)                                                                                       

In [5]: I = Identity(2)                                                                                           

In [6]: assemble(inner(f, I)*dx).dat.data                                                                         
Out[6]: 
array([[0.5, 0. , 0.5],
       [0.5, 0. , 0.5]])

In [13]: V = TensorFunctionSpace(mesh, "DG", 1, symmetry=True)                                                    

In [14]: f = TestFunction(V)                                                                                      

In [15]: I = Identity(2)                                                                                          

In [16]: assemble(inner(f, I)*dx).dat.data                                                                        
Out[16]: 
array([[0.16666667, 0.        , 0.16666667],
       [0.16666667, 0.        , 0.16666667],
       [0.16666667, 0.        , 0.16666667],
       [0.16666667, 0.        , 0.16666667],
       [0.16666667, 0.        , 0.16666667],
       [0.16666667, 0.        , 0.16666667]])
```

**Comment by pbrubeck** [comment]
This issue has been addressed within #1963
